consensus/src/main/java/org/ethereum/beacon/consensus/SpecHelpers.java

```diff
@@ -1194,6 +1197,114 @@ public void checkShardRange(UInt64 shard) {
   }
 
+  // def lmd_ghost(store, start):
+  //    validators = start.state.validator_registry
+  //    active_validators = [validators[i] for i in
+  //                         get_active_validator_indices(validators, start.state.slot)]
+  //    attestation_targets = [get_latest_attestation_target(store, validator)
+  //                           for validator in active_validators]
+  //    def get_vote_count(block):
+  //        return len([target for target in attestation_targets if
+  //                    get_ancestor(store, target, block.slot) == block])
+  //
+  //    head = start
+  //    while 1:
+  //        children = get_children(head)
+  //        if len(children) == 0:
+  //            return head
+  //        head = max(children, key=get_vote_count)
+  public BeaconBlock lmd_ghost(
+      BeaconBlock startBlock,
+      BeaconState state,
+      Function<Hash32, Optional<BeaconBlock>> getBlock,
+      Function<Hash32, List<BeaconBlock>> getChildrenBlocks,
+      Function<ValidatorRecord, Attestation> get_latest_attestation) {
+    List<ValidatorRecord> validators = state.getValidatorRegistry();
+    List<UInt24> active_validator_indices =
+        get_active_validator_indices(validators, state.getSlot());
+
+    List<ValidatorRecord> active_validators = new ArrayList<>();
+    for (UInt24 index : active_validator_indices) {
+      active_validators.add(validators.get(index.getValue()));
+    }
+
+    List<BeaconBlock> attestation_targets = new ArrayList<>();
+    for (ValidatorRecord validatorRecord : active_validators) {
+      attestation_targets.add(
+          get_latest_attestation_target(validatorRecord, get_latest_attestation, getBlock));
+    }
+
+    BeaconBlock head = startBlock;
+    while (true) {
+      List<BeaconBlock> children = getChildrenBlocks.apply(head.getHash());
+      if (children.isEmpty()) {
+        return head;
+      } else {
+        head =
+            children.stream()
+                .max(Comparator.comparingInt(o -> get_vote_count(o, attestation_targets, getBlock)))
+                .orElseThrow(() -> new RuntimeException("Couldn't find maximum voted block"));
+      }
+    }
+  }
+
+  /**
+   * Let get_latest_attestation_target(store, validator) be the target block in the attestation
+   * get_latest_attestation(store, validator).
+   *
+   * @param get_latest_attestation Let get_latest_attestation(store, validator) be the attestation
+   *     with the highest slot number in store from validator. If several such attestations exist,
+   *     use the one the validator v observed first.
+   */
+  private BeaconBlock get_latest_attestation_target(
+      ValidatorRecord validatorRecord,
+      Function<ValidatorRecord, Attestation> get_latest_attestation,
+      Function<Hash32, Optional<BeaconBlock>> getBlock) {
+    Attestation latest = get_latest_attestation.apply(validatorRecord);
+    return getBlock
+        .apply(latest.getData().getJustifiedBlockRoot())
+        .orElseThrow(
```

Comment:
I don't think this is exceptional case. There could be a recent network attestation which refers to not yet imported block, rare case for justified block, though. I think in this case an attestation for prior slot should be picked.